From 4690808e65f1abda5e779620672da14c830aaa89 Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Wed, 17 May 2017 12:53:22 +0200
Subject: use transientparent directly

Summary:
don't parent directly as we don't want unwanted deletions.
ensure we have a menu, not parented yet, the container exists,
both the menu and the container's toplevel widget QWindows exist,
then set the transient parent of the qmenu's qwindow to the
container's toplevel widget's qwindow. this fixes positioning in
wayland as the protocol requires every popup menu having a transient parent

Test Plan:
okular doesn't crash anymore, konsole's bookmarks menu still
correctly positioned in wayland

Reviewers: lbeltrame

Reviewed By: lbeltrame

Subscribers: lbeltrame, #frameworks

Tags: #frameworks

Differential Revision: https://phabricator.kde.org/D5900
---
 src/kxmlguifactory_p.cpp | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/src/kxmlguifactory_p.cpp b/src/kxmlguifactory_p.cpp
index eb28cb2..c567559 100644
--- a/src/kxmlguifactory_p.cpp
+++ b/src/kxmlguifactory_p.cpp
@@ -26,6 +26,7 @@
 #include <QWidget>
 #include <QDebug>
 #include <QMenu>
+#include <QWindow>
 
 #include "debug.h"
 #include <assert.h>
@@ -593,8 +594,18 @@ bool BuildHelper::processActionElement(const QDomElement &e, int idx)
 
     parentNode->container->insertAction(before, action);
 
-    if (action->menu() && !action->menu()->parentWidget()) {
-        action->menu()->setParent(parentNode->container, action->menu()->windowFlags());
+    //for wayland:
+    //ensure we have a menu, not parented yet, the container exists,
+    //both the menu and the container's toplevel widget QWindows exist,
+    //then set the transient parent of the qmenu's qwindow to the
+    //container's toplevel widget's qwindow. this fixes positioning in
+    //wayland as the protocol requires every popup menu having a transient parent
+    if (action->menu() && !action->menu()->parentWidget() &&
+        parentNode->container && parentNode->container->window() &&
+        action->menu()->winId() &&
+        parentNode->container->window()->winId()) {
+        action->menu()->windowHandle()->setTransientParent(parentNode->container->window()->windowHandle());
+
     }
     // save a reference to the plugged action, in order to properly unplug it afterwards.
     containerClient->actions.append(action);
-- 
cgit v0.11.2
